### PR TITLE
Backup & restore system

### PR DESCRIPTION
**Pull request by Spitfireap**
Description of the issue/feature this PR addresses:
Having a way to back up all the data needed to restore modoboa. It is possible to disable mail backup and it also backup some config file (amavis and postwhite)

Current behavior before PR:
Backup and restore system working. It needs test !! Backup is stable, however restoring require a long time to test

### CODE REVIEW

**Comment by Spitfireap** [comment]
I need people testing this, feel free to do so !

**Comment by ionescu77** [comment]
This would be a nice functionality. I'll have a look. Thx!

Your message on Google groups was informative (not much activity there tough). I might paste it here.

**Comment by ionescu77** [comment]
Copy paste from Google groups

Hi everyone !
 
 
I've developed a small addition to modoboa-installer to be able to 
backup and restore a modoboa instance installed WITH the installer and 
restore it on a FRESH AND NEW machine. There are multiple options 
available for now like discarding mail backup (when you have too much 
mails for instance).
 
It requires you to have the original `installer.cfg` used to deploy your 
current instance. If you don't have it anymore you can also run the 
utility with `--stop-after-configfile-check` and edit `installer.cfg` so it 
reflect your current instance (most importantly databases password !!).
 
 
For the backup side you have the option to run it interactively (simply 
use `--backup`), not interactively by providing a path (--backup --bash 
PATH) or non interactively without providing a path, with a default 
location (--backup --sbash). And you can also discard mail backup  
(`--backup --no-mails`).
 
 
For the restore side you simply need to use `--restore PATH_TO_BACKUP_DIR`.
 
 
I would be glad if you can test it if you have a spare vm/vps, or simply 
need to transfer your instance to a new rig.
 
To do it :

```

git clone https://github.com/Spitfireap/modoboa-installer.git
# don't do it in same folder that contains modoboa-installer !
 

cd modoboa-installer
 
git checkout restore
 
cp OLD_INSTALLER_CFG ./
# if you have it
 
python3 run.py --backup YOUR_DOMAIN
# or anything else than 
backup if you need to do restore for example
 
```
 
Any comments, ideas, or bug reports are appreciated !
 
you can find the PR here : 
https://github.com/modoboa/modoboa-installer/pull/424
 
 
Best regards,
 
Adrien

**Comment by angeloluidens** [comment]
Thank you very much, Adrien! (repeated from Google Groups)

Will start testing it.

Best regards,
Angelo

**Comment by Spitfireap** [comment]
Thanks for copying it, and thanks for giving it a try :)

File: modoboa_installer/utils.py [hunk 1]
```diff
@@ -163,19 +170,32 @@ def copy_from_template(template, dest, context):
...
     if os.path.exists(dest):
-        return
+        return isPresent
     if upgrade:
         printcolor(
             "You cannot upgrade an existing installation without a "
             "configuration file.", RED)
         sys.exit(1)
+    elif backup:
+        isPresent = False
```
Comment (tonioo) [inline modoboa_installer/utils.py:184]: is_present

File: run.py [hunk 2]
```diff
@@ -72,16 +91,48 @@ def main(input_args):
...
+    if args.restore is not None:
+        is_restoring = True
+        args.configfile = os.path.join(args.restore, "installer.cfg")
+        if not os.path.exists(args.configfile):
+            utils.printcolor("installer.cfg from backup not found!", utils.RED)
+            sys.exit(1)
+
     utils.printcolor("Welcome to Modoboa installer!\n", utils.GREEN)
-    utils.check_config_file(args.configfile, args.interactive, args.upgrade)
+    is_config_file_availible = utils.check_config_file(
```
Comment (tonioo) [inline run.py:126]: is_config_file_available

File: modoboa_installer/scripts/backup.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+        if path_exists and os.path.isfile(path):
+            utils.printcolor(
+                "Error, you provided a file instead of a directory!", utils.RED)
+            return False
+
+        if not path_exists:
+            if not self.silent_backup:
+                create_dir = input(
+                    f"\"{path}\" doesn't exists, would you like to create it ? [Y/n]\n").lower()
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:53]: doesn't exist, would you like to create it? (no space before interrogation mark in english)

File: modoboa_installer/scripts/backup.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+            if self.silent_backup or (not self.silent_backup and (create_dir == "y" or create_dir == "yes")):
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:55]: create_dir.startswith("y")

File: modoboa_installer/scripts/backup.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                pw = pwd.getpwnam("root")
+                utils.mkdir_safe(path, stat.S_IRWXU |
+                                 stat.S_IRWXG, pw[2], pw[3])
+            else:
+                utils.printcolor(
+                    "Error, backup dir not present.", utils.RED
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:61]: directory

File: modoboa_installer/scripts/backup.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                )
+                return False
+
+        if len(os.listdir(path)) != 0:
+            if not self.silent_backup:
+                delete_dir = input(
+                    "Warning : backup folder is not empty, it will be purged if you continue... [Y/n]\n").lower()
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:68]: We should be consistent about messages (sometimes you talk about directories and sometimes about folders). Please choose one

File: modoboa_installer/scripts/backup.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        return True
+
+    def set_path(self):
+        """Setup backup directory."""
+        if self.silent_backup:
+            if self.backup_path is None:
+                if self.config.has_option("backup", "default_path"):
+                    path = self.config.get("backup", "default_path")
+                else:
+                    path = f"./modoboa_backup/"
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:102]: Can't you use the value indicated in config_dict_template?

File: modoboa_installer/scripts/backup.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            else:
+                if not self.validate_path(self.backup_path):
+                    utils.printcolor(
+                        f"Path provided : {self.backup_path}", utils.BLUE)
+                    sys.exit(1)
+        else:
+            user_value = None
+            while user_value == "" or user_value is None or not self.validate_path(user_value):
+                utils.printcolor(
+                    "Enter backup path, please provide an empty folder.", utils.MAGENTA)
```
Comment (tonioo) [inline modoboa_installer/scripts/backup.py:115]: Enter backup path (it must be an empty folder)

**Comment by Spitfireap** [comment]
Thank you for the review, it should be good now :)

**Pull request merged**